### PR TITLE
Fixed for #725, #150

### PR DESCRIPTION
**Pull request by yosukehara**
I've fixed both #725 and #150 which adhere to the following sequence diagram:

leofs-del-dir-seq-diagram

I've implemented `leofs-adm delete-bucket-stats [<BUCKET_NAME>]` so that we can easily check a current state of each delete-bucket processing.

```
$ ./leofs-adm delete-bucket-stats test_1
node                         | node             | state
-----------------------------+------------------+-----------------------------
storage_0@127.0.0.1          | monitoring       | 2017-06-29 11:54:51 +0900
storage_1@127.0.0.1          | monitoring       | 2017-06-29 11:54:52 +0900
storage_2@127.0.0.1          | finished         | 2017-06-29 11:54:47 +0900
storage_3@127.0.0.1          | finished         | 2017-06-29 11:54:52 +0900

$ ./leofs-adm delete-bucket-stats
- Bucket: test_1
node                         | node             | state
-----------------------------+------------------+-----------------------------
storage_0@127.0.0.1          | pending          | 2017-06-29 11:54:05 +0900
storage_1@127.0.0.1          | pending          | 2017-06-29 11:54:05 +0900
storage_2@127.0.0.1          | monitoring       | 2017-06-29 11:54:20 +0900
storage_3@127.0.0.1          | pending          | 2017-06-29 11:54:05 +0900


- Bucket: test_10
node                         | node             | state
-----------------------------+------------------+-----------------------------
storage_0@127.0.0.1          | pending          | 2017-06-29 11:54:13 +0900
storage_1@127.0.0.1          | pending          | 2017-06-29 11:54:13 +0900
storage_2@127.0.0.1          | pending          | 2017-06-29 11:54:13 +0900
storage_3@127.0.0.1          | pending          | 2017-06-29 11:54:13 +0900


- Bucket: test_11
node                         | node             | state
-----------------------------+------------------+-----------------------------
storage_0@127.0.0.1          | pending          | 2017-06-29 11:54:14 +0900
storage_1@127.0.0.1          | pending          | 2017-06-29 11:54:14 +0900
storage_2@127.0.0.1          | pending          | 2017-06-29 11:54:14 +0900
storage_3@127.0.0.1          | pending          | 2017-06-29 11:54:14 +0900


- Bucket: test_12
node                         | node             | state
-----------------------------+------------------+-----------------------------
storage_0@127.0.0.1          | pending          | 2017-06-29 11:54:14 +0900
storage_1@127.0.0.1          | pending          | 2017-06-29 11:54:14 +0900
storage_2@127.0.0.1          | pending          | 2017-06-29 11:54:14 +0900
storage_3@127.0.0.1          | pending          | 2017-06-29 11:54:14 +0900


- Bucket: test_2
node                         | node             | state
-----------------------------+------------------+-----------------------------
storage_0@127.0.0.1          | finished         | 2017-06-29 11:54:29 +0900
storage_1@127.0.0.1          | monitoring       | 2017-06-29 11:54:28 +0900
storage_2@127.0.0.1          | monitoring       | 2017-06-29 11:54:29 +0900
storage_3@127.0.0.1          | monitoring       | 2017-06-29 11:54:31 +0900

```

### CODE REVIEW

File: apps/leo_storage/src/leo_storage_handler_del_directory.erl [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+           MQId, Directory) of
+        {ok,_TotalMsgs} ->
+            erlang:send(From, {enqueued, MQId, Type, Directory});
+        {error, Cause} ->
+            ?error("insert_messages/3", [{cause, Cause}]),
+            erlang:send(From, {failed, MQId, Type, Directory})
+    end.
+
+
+%% @doc Retrieve available wokers (not ongoingg and monitoring workers)
```
Comment (mocchira) [inline apps/leo_storage/src/leo_storage_handler_del_directory.erl]: The comment (not ongoing and monitoring workers) is different from what actually implemented in the function (pending/ongoing/monitoring state records are returned).

File: apps/leo_storage/src/leo_storage_handler_del_directory.erl [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+
+%%====================================================================
+%% Private Functions
+%%====================================================================
+%% @private
+-spec(dequeue(State) ->
+             {ok, State} when State::#state{}).
+dequeue(#state{num_of_workers = NumOfWorkers} = State) ->
+    case get_ongoing_workers(NumOfWorkers) of
+        {ok, DelBucketStateList} ->
```
Comment (mocchira) [inline apps/leo_storage/src/leo_storage_handler_del_directory.erl]: Maybe commenting about what the below listing does (retrieving available mq_id) would be good for those who try to read the code for the first time.

File: apps/leo_manager/include/leo_manager.hrl [hunk 3]
```diff
@@ -404,9 +429,77 @@
...
           node :: atom(),
           rebalance_info = [] :: list(tuple()),
-          timestamp = 0 :: pos_integer()
+          timestamp = 1 :: pos_integer()
+         }).
+
+
+-define(STATE_INVALID, 0).
+-define(STATE_PENDING, 1).
+-define(STATE_ONGOING, 2).
```
Comment (mocchira) [inline apps/leo_manager/include/leo_manager.hrl]: Maybe adding a comment about what STATE_ONGOING (Finding objects to be deleted) means would be good for maintainers.

File: apps/leo_manager/include/leo_manager.hrl [hunk 4]
```diff
@@ -404,9 +429,77 @@
...
+-define(STATE_MONITORING, 3).
```
Comment (mocchira) [inline apps/leo_manager/include/leo_manager.hrl]: Maybe adding a comment about what STATE_MONITORING (Check the state of the deletion bucket) means would be good for maintainers.

**Comment by mocchira** [comment]
Also confirmed this PR covered all the design considerations described here https://github.com/leo-project/leofs/issues/725#issuecomment-302606104 other than the last one(this one is out of the scope).

**Comment by mocchira** [comment]
@yosukehara noticed the below patch is needed for leofs-adm.
```diff
diff --git a/leofs-adm b/leofs-adm                                                                       │tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN
index 0bbc0da..47ce368 100755                                                                            │tcp        0      0 0.0.0.0:4369            0.0.0.0:*               LISTEN
--- a/leofs-adm                                                                                          │tcp        0      0 0.0.0.0:43699           0.0.0.0:*               LISTEN
+++ b/leofs-adm                                                                                          │tcp        0      0 0.0.0.0:13075           0.0.0.0:*               LISTEN
@@ -959,6 +959,7 @@ usage() {                                                                            │tcp        0      0 0.0.0.0:13076           0.0.0.0:*               LISTEN
     output ""                                                                                           │tcp        0      0 0.0.0.0:13078           0.0.0.0:*               LISTEN
     usage_add_bucket min                                                                                │tcp        0      0 0.0.0.0:33558           0.0.0.0:*               LISTEN
     usage_delete_bucket min                                                                             │tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN
+    usage_delete_bucket_stats min                                                                       │tcp        0      0 0.0.0.0:13079           0.0.0.0:*               LISTEN
     usage_get_bucket min                                                                                │tcp        0      0 0.0.0.0:13080           0.0.0.0:*               LISTEN
     usage_get_buckets min                                                                               │tcp        0      0 0.0.0.0:10010           0.0.0.0:*               LISTEN
     usage_chown_bucket min
```

Now issuing leofs-adm without any argument doesn't show the delete-bucket-stats help.

**Comment by mocchira** [comment]
@yosukehara Just in case, I'd like to share that now `./test/test_it_del_bucket_while_one_node_is_unavailable.sh` eventually get to show `Delete-bucket's stats not found`, however delete-bucket-stats should keep showing the stats while there is still  a **pending** target (storage_0).

This problem will be solved when this review comment https://github.com/leo-project/leofs/pull/781#discussion_r124718305 is taken into account.

**Comment by yosukehara** [comment]
@mocchira I've fixed those issues, please review them, again.

File: apps/leo_manager/src/leo_manager_api.erl [hunk 5]
```diff
@@ -2580,3 +2598,14 @@ call_remote_node_fun_1([#node_state{node = Node,
     end;
 call_remote_node_fun_1([_|Rest], NodeStr, Method, Args) ->
     call_remote_node_fun_1(Rest, NodeStr, Method, Args).
+
+
+%% Notification of a del-bucket's state
+-spec(notify_del_dir_state(Node, BucketName, State) ->
+             ok | {error, Cause} when Node::node(),
+                                      BucketName::binary(),
+                                      State::pending|ongoing|monitoring|finished,
```
Comment (mocchira) [inline apps/leo_manager/src/leo_manager_api.erl]: s/ongoing/enqueuing/

File: apps/leo_storage/src/leo_storage_handler_del_directory.erl [hunk 6]
```diff
@@ -0,0 +1,646 @@
...
+           MQId, Directory) of
+        {ok,_TotalMsgs} ->
+            erlang:send(From, {enqueued, MQId, Type, Directory});
+        {error, Cause} ->
+            ?error("insert_messages/3", [{cause, Cause}]),
+            erlang:send(From, {failed, MQId, Type, Directory})
+    end.
+
+
+%% @doc Retrieve the records whose state is pending/ongoing/monitoring.
```
Comment (mocchira) [inline apps/leo_storage/src/leo_storage_handler_del_directory.erl]: s/ongoing/enqueuing/

**Comment by yosukehara** [comment]
@mocchira I've fixed the spec of `leo_manager_api:notify_del_dir_state/3` and the doc of `leo_storage_handler_del_directory:get_ongoing_workers/1`.

**Comment by mocchira** [comment]
LGTM.

**Pull request merged**